src/datahike/transactor.cljc

```diff
@@ -0,0 +1,54 @@
+(ns datahike.transactor
+  (:require [datahike.core :as d]
+            [superv.async :refer [<?? <??- S thread-try]]
+            [taoensso.timbre :as log]
+            [clojure.core.async :refer [>!! chan close! promise-chan]]))
+
+(defprotocol PTransactor
+  ; Send a transaction. Returns a channel that resolves when the transaction finalizes.
+  (send-transaction! [_ tx-data tx-fn])
+  ; Returns a channel that resolves when the transactor has shut down.
+  (shutdown [_]))
+
+(defrecord LocalTransactor
+  [rx-queue rx-thread]
+  PTransactor
+  (send-transaction! [_ tx-data tx-fn]
+    (let [p (promise-chan)]
+      (>!! rx-queue {:tx-data tx-data :callback p :tx-fn tx-fn})
+      p))
+
+  (shutdown [_]
+    (close! rx-queue)
+    rx-thread))
+
+(defn create-rx-thread
+  [connection rx-queue update-and-flush-db]
+  (thread-try
+    S
+    (let [resolve-fn (memoize resolve)]
+      (loop []
+        (if-let [{:keys [tx-data callback tx-fn]} (<??- rx-queue)]
+          (do
+            (let [update-fn (resolve-fn tx-fn)
+                  tx-report (try (update-and-flush-db connection tx-data update-fn)
+                                 ; Only catch ExceptionInfo here (intentionally rejected transactions).
+                                 ; Any other exceptions should crash the transactor and signal the supervisor.
+                                 (catch clojure.lang.ExceptionInfo e e))]
+              (when (some? callback)
+                (>!! callback tx-report)))
```

Comment:
I would also consider not blocking the transactor on calling back the listening processes, although this was done for Datascript in memory before. Otherwise slow callbacks will jam the transactor. Using `put!` should be fine here, right? This would also render `thread-try` and `<??-` redundant, we can just use a go-loop and run this logic fully asynchronously. As a side-note, the transaction logic here could batch transactions and only flush at the end (i.e. doing updates until the rx-queue is empty and then flush and then notify all sending processes).

Reply:
This batching logic could be shared between transactors.

Reply:
`>!!` won't block if the buffer has space (promise chan has 1 slot so it wont be full because nothing else is putting onto it) however I will change it to use `put!` here anyways since >!! uses put! under the hood so it's just less code to execute.

Reply:
> This would also render thread-try and <??- redundant, we can just use a go-loop and run this logic fully asynchronously.

Can't do that because the loop is io bound on the update-and-flush. You're not supposed to do significant IO on the go thread pool. This also keeps the CPU cache "hot" on the main loop here if it matters at all (it might under real stress) - less context switching.

> batch transactions and only flush at the end (i.e. doing updates until the rx-queue is empty and then flush and then notify all sending processes).

I'm sure something like this is possible but also needs to be carefully implemented. Like a max timeout/# of transactions before a flush is forced. I would save this for a follow up PR as a performance optimization. We don't want callbacks to be blocked for a long time if the queue is long for example, because that could lead to a breakdown in performance under load. There is probably a 'sweet spot' to hit on these parameters.